x/poolmanager/client/query_proto_wrap.go

```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
+	}
+
+	// If ExternalPrice is specified we need to adjust the maxPriceImpact based on the deviation between spot and
+	// external price.
+	adjustedMaxPriceImpact := req.MaxPriceImpact
+	if !req.ExternalPrice.IsZero() {
+		priceDeviation := spotPrice.Sub(req.ExternalPrice).Quo(req.ExternalPrice)
+		adjustedMaxPriceImpact = adjustedMaxPriceImpact.Sub(priceDeviation)
+
+		// If the adjusted max price impact is negative or zero it means the difference between spot and external
+		// already exceeds the max price impact.
+		if adjustedMaxPriceImpact.IsZero() || adjustedMaxPriceImpact.IsNegative() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+			}, nil
+		}
+	}
+
+	// Process the estimates according to the pool type.
+	switch poolI.GetType() {
+	case types.Balancer:
+		return q.estimateTradeBasedOnPriceImpactBalancerPool(
+			ctx, req, spotPrice, adjustedMaxPriceImpact, swapModule, poolI,
+		)
+	case types.Stableswap:
+		return q.estimateTradeBasedOnPriceImpactStableSwapPool(
+			ctx, req, spotPrice, adjustedMaxPriceImpact, swapModule, poolI,
+		)
+	case types.Concentrated:
+		return q.estimateTradeBasedOnPriceImpactConcentratedLiquidity(
+			ctx, req, spotPrice, adjustedMaxPriceImpact, swapModule, poolI,
+		)
+	default:
+		return nil, status.Error(codes.Internal, "pool type not supported")
+	}
+}
+
+// estimateTradeBasedOnPriceImpactBalancerPool estimates a trade based on price impact for a balancer pool type.
+// For a balancer pool if an amount entered is greater than the total pool liquidity the trade estimated would be
+// the full liquidity of the other token. If the amount is small it would return a close 1:1 trade of the
+// smallest units.
+func (q Querier) estimateTradeBasedOnPriceImpactBalancerPool(
+	ctx sdk.Context,
+	req queryproto.EstimateTradeBasedOnPriceImpactRequest,
+	spotPrice, adjustedMaxPriceImpact sdk.Dec,
+	swapModule types.PoolModuleI,
+	poolI types.PoolI,
+) (*queryproto.EstimateTradeBasedOnPriceImpactResponse, error) {
+	// There isn't a case where the tokenOut could be zero or an error is received but those possibilities are handled
+	// anyway.
+	tokenOut, err := swapModule.CalcOutAmtGivenIn(ctx, poolI, req.FromCoin, req.ToCoinDenom, sdk.ZeroDec())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if tokenOut.IsZero() {
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+			OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+		}, nil
+	}
+
+	// Validate if the trade as is respects the price impact, if it does re-estimate it with a swap fee and return
+	// the result.
+	currTradePrice := sdk.NewDec(req.FromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+	priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
```

Comment:
Whats the reason we take abs value here? can any of these value be negative in the first place?

Reply:
I haven't found a value that can be negative, I can remove it but I am not sure if there is a pool configuration that could give more tokens than inputted especially in the balancer pool type and how its weights work.

Reply:
Ah I see your concern here, let's put a comment here explaining the Abs we take here is a saftey measure